fix(#48173): correct api status code

ganeshdipdumbare opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
- fixes #48173
- return correct status code for invalid params in the api

**- How I did it**
- added error wrapper to return correct error status code

**- How to verify it**
- already used at some places, missed at some places

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
- return correct status code for invalid params in the api
```

**- A picture of a cute animal (not mandatory but encouraged)**

ganeshdipdumbare committed:
fix inconsistent api status codes

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -129,7 +129,7 @@ func (ir *imageRouter) postImagesCreate(ctx context.Context, w http.ResponseWrit
 			resp, err := remotecontext.GetWithStatusError(u.String())
 			if err != nil {
-				return err
+				return errdefs.InvalidParameter(err)
```
I think `GetWithStatusError` above already is used to set the right status; we should check if that's always the error-type we want to return, but we probably shouldn't always consider this an invalid parameter https://github.com/moby/moby/blob/980d8e5d54d8df31ba0e75e1d9dd474d9d91308a/builder/remotecontext/remote.go#L46-L78

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -175,7 +175,7 @@ func (ir *imageRouter) postImagesPush(ctx context.Context, w http.ResponseWriter
 		var err error
 		authConfig, err = registry.DecodeAuthConfigBody(r.Body)
 		if err != nil {
-			return errors.Wrap(err, "bad parameters and missing X-Registry-Auth")
```
DecodeAuthConfigBody above already returns a typed error; even if wrapped , the API should be able to use that, so changing here should not be needed https://github.com/moby/moby/blob/980d8e5d54d8df31ba0e75e1d9dd474d9d91308a/api/types/registry/authconfig.go#L73-L82

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -215,15 +215,15 @@ func (ir *imageRouter) postImagesPush(ctx context.Context, w http.ResponseWriter
 			p, err := httputils.DecodePlatform(formPlatform)
 			if err != nil {
-				return err
+				return errdefs.InvalidParameter(err)
```
DecodePlatform above already returns a typed error, so this should not be needed https://github.com/moby/moby/blob/980d8e5d54d8df31ba0e75e1d9dd474d9d91308a/api/server/httputils/form.go#L119

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -215,15 +215,15 @@ func (ir *imageRouter) postImagesPush(ctx context.Context, w http.ResponseWriter
 	if err := ir.backend.PushImage(ctx, ref, platform, metaHeaders, authConfig, output); err != nil {
 		if !output.Flushed() {
-			return err
+			return errdefs.InvalidParameter(err)
```
it looks like the backend PushImage already returns a typed error, and we can't assume it's always an invalid parameter here, so this should probably not be changed https://github.com/moby/moby/blob/980d8e5d54d8df31ba0e75e1d9dd474d9d91308a/daemon/images/image_push.go#L21

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -472,12 +472,12 @@ func (ir *imageRouter) getImagesSearch(ctx context.Context, w http.ResponseWrite
 	pruneFilters, err := filters.FromJSON(r.Form.Get("filters"))
 	if err != nil {
-		return err
+		return errdefs.InvalidParameter(err)
```
filters.FromJSON already returns typed errors, so this should probably not be changed https://github.com/moby/moby/blob/980d8e5d54d8df31ba0e75e1d9dd474d9d91308a/api/types/filters/parse.go#L85

thaJeztah reviewed (changes requested):
Looks like quite some of these should be checked, and may not be needed (as they would be masking already typed errors); I left some comments on those, but haven't checked all of them

ganeshdipdumbare committed:
add correct status code and update swagger doc for apis

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>

ganeshdipdumbare commented:
> Looks like quite some of these should be checked, and may not be needed (as they would be masking already typed errors); I left some comments on those, but haven't checked all of them

You are right. I have again checked the the issue which was opened for the same. I have added correct typed errors now for given apis in the issue and also added the same in the doc.

ganeshdipdumbare committed:
fix: remove unwanted error typing

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>

ganeshdipdumbare committed:
Trigger pipeline

ganeshdipdumbare committed:
Trigger pipeline

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>

ganeshdipdumbare committed:
fix: trigger pipeline

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>

ganeshdipdumbare committed:
fix: trigger pipeline

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>

thaJeztah commented:
Oh! yes, there's some regression we identified in a  dependency; that's on our radar and being looked into in upstream.

Generally, we rebase and squash commits; also happy to restart CI on failures for you.

thaJeztah commented:
Oh! I see I had a review pending that I didn't submit; hold on

ganeshdipdumbare committed:
fix: add correct error typing

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>

ganeshdipdumbare commented on api/server/router/image/image_routes.go:
```diff
@@ -248,7 +248,7 @@ func (ir *imageRouter) getImagesGet(ctx context.Context, w http.ResponseWriter,
 	if err := ir.backend.ExportImage(ctx, names, output); err != nil {
 		if !output.Flushed() {
-			return err
+			return errdefs.InvalidParameter(err)
```
Yup. I guess it is much more better to implement typed errors in the pkg `tarexport`.
I have reverted changes from here and create a ticket for implementing the above mentioned change - https://github.com/moby/moby/issues/48408

ganeshdipdumbare commented on api/server/router/image/image_routes.go:
```diff
@@ -95,7 +95,7 @@ func (ir *imageRouter) postImagesCreate(ctx context.Context, w http.ResponseWrit
 
 		if err := validateRepoName(ref); err != nil {
-			return err
+			return errdefs.Forbidden(err)
```
Ah got it. Fixed it by adding suggested changes.

vvoland replied:
Ah, related: https://github.com/moby/moby/pull/48386#discussion_r1741235344

ganeshdipdumbare commented on api/server/router/plugin/plugin_routes.go:
```diff
@@ -69,7 +70,7 @@ func (pr *pluginRouter) getPrivileges(ctx context.Context, w http.ResponseWriter
 	ref, _, err := parseRemoteRef(r.FormValue("remote"))
 	if err != nil {
-		return err
+		return errdefs.InvalidParameter(err)
```
Sounds good. I have added above mentioned changes now.

vvoland commented on api/server/router/image/image_routes.go:
```diff
@@ -95,7 +95,7 @@ func (ir *imageRouter) postImagesCreate(ctx context.Context, w http.ResponseWrit
 
 		if err := validateRepoName(ref); err != nil {
-			return errdefs.Forbidden(err)
+			return errdefs.InvalidParameter(err)
```
Note: The reason behind this one being `errdefs.Forbidden` is that this is replicating the error from the graphdrivers (for the containerd implementation)
https://github.com/moby/moby/blob/095d2a29a3840815704c9c8f3c29fc755e36a40b/distribution/pull.go#L86
https://github.com/moby/moby/blob/62b33a2604c0bacdd26bfbe7303cb6e2ed26d432/distribution/errors.go#L208-L214

vvoland replied:
Continuing here: https://github.com/moby/moby/pull/48386#discussion_r1741235344

vvoland commented on api/server/router/image/image_routes.go:
```diff
@@ -95,7 +95,7 @@ func (ir *imageRouter) postImagesCreate(ctx context.Context, w http.ResponseWrit
 		}
 
 		if err := validateRepoName(ref); err != nil {
-			return errdefs.Forbidden(err)
```
If we want to change `Forbidden` into `InvalidParameter` we should also do it for this error:
https://github.com/moby/moby/blob/62b33a2604c0bacdd26bfbe7303cb6e2ed26d432/distribution/errors.go#L208-L214

ganeshdipdumbare replied:
changed.

ganeshdipdumbare committed:
change reservedNameError to invalid param

Signed-off-by: Ganeshdip Dumbare <ganeshdip.dumbare@gmail.com>
